hydra-node/test/Hydra/Chain/Direct/TxSpec.hs

```diff
@@ -1,24 +1,87 @@
+{-# LANGUAGE TypeApplications #-}
+
 -- | Unit tests for our "hand-rolled" transactions as they are used in the
 -- "direct" chain component.
 module Hydra.Chain.Direct.TxSpec where
 
 import Hydra.Prelude
 import Test.Hydra.Prelude
 
-import Cardano.Api (
-  makeTransactionBody,
- )
-import Gen.Cardano.Api.Typed (genTxIn)
-import Hydra.Chain.Direct.Tx (initTx)
-import Test.QuickCheck (counterexample, forAll, property)
-import Test.QuickCheck.Hedgehog (hedgehog)
+import Cardano.Binary (serialize)
+import Cardano.Ledger.Alonzo.Data (Data (Data))
+import Cardano.Ledger.Alonzo.Tx (ValidatedTx (ValidatedTx, wits, body), outputs)
+import Cardano.Ledger.Alonzo.TxWitness (TxWitness (txdats), nullDats, unTxDats)
+import qualified Data.ByteString.Lazy as LBS
+import qualified Data.Map as Map
+import Hydra.Chain (HeadParameters (..), toOnChainTx, PostChainTx (InitTx))
+import Hydra.Chain.Direct.Tx (initTx, observeTx, constructTx)
+import Hydra.Contract.ContestationPeriod (contestationPeriodFromDiffTime)
+import Hydra.Contract.Head (State (Initial))
+import Hydra.Contract.Party (partyFromVerKey)
+import Hydra.Party (vkey)
+import Plutus.V1.Ledger.Api (toBuiltinData, toData)
+import Test.Cardano.Ledger.Alonzo.Serialisation.Generators ()
+import Test.QuickCheck (counterexample, (===), (==>))
+import Hydra.Ledger.Simple (SimpleTx)
+import Hydra.Chain.Direct (Era)
+import Cardano.Ledger.Mary.Value (Value(Value), PolicyID, AssetName)
+import Cardano.Ledger.Alonzo (TxOut)
+import Cardano.Ledger.Alonzo.TxBody (TxOut(TxOut))
+import Cardano.Ledger.Crypto (StandardCrypto)
 
 spec :: Spec
 spec =
-  parallel $
+  parallel $ do
+    prop "observeTx . constructTx roundtrip" $ \postTx txIn time ->
+      isImplemented postTx ==> -- TODO(SN): test all constructors
+      observeTx (constructTx txIn postTx) === Just (toOnChainTx @SimpleTx time postTx)
+
     describe "initTx" $ do
-      prop "can be used to construct a non-balanced TxBody" $ \params ->
-        forAll (hedgehog genTxIn) $ \txIn ->
-          case makeTransactionBody $ initTx params txIn of
-            Left err -> counterexample ("TxBodyError: " <> show err) False
-            Right _ -> property True
+      prop "can construct & serialize unsigned initTx" $ \txIn params ->
+        let tx = initTx params txIn
+            cbor = serialize tx
+            len = LBS.length cbor
+         in counterexample ("Tx: " <> show tx) $
+              counterexample ("Tx serialized size: " <> show len) $
+                len < 16000 -- TODO(SN): use real max tx size
+
+      prop "contains some datums" $ \txIn params ->
+        let ValidatedTx{wits} = initTx params txIn
+            dats = txdats wits
+         in counterexample ("TxDats: " <> show dats) $
+              not $ nullDats dats
```

Comment:
Do we really need a `prop` here? AFAIK, `txIn` and `params` are not used to define `wits`.

Reply:
These props are really temporary (I guess).. they are artifacts of me piecing together some Tx

Reply:
Also I was lazy in finding values of `txIn` and `params` so Quickcheck was giving me some :)